docs/decisions/0002-build-time-customization-with-npm-aliases.rst

```diff
@@ -0,0 +1,72 @@
+2. Build time customization using NPM aliases
+---------------------------------------------
+
+Status
+------
+
+Accepted
+
+Context
+-------
+
+Frontend applications created throughout FY2019 contain hardcoded edX brand specific elements
+such as the site's header, footer, logo, visual style, and navigational links. This enabled
+teams to move more quickly in efforts to adopt a micro-frontend architecture to enable more
+rapid UI innovation in the future. There was no easy path for these new applications to be
+incorporated into the Open edX platform where they need to be branded properly.
+
+Decision
+--------
+
+In order to make frontend applications brand agnostic, we will split branded elements into
+npm packages such as ``frontend-component-header``. Frontend applications will expect to
+find components or other exports according to a defined interface. Package interfaces will
+be defined in the README for each npm package repository.
+
+.. code-block:: javascript
+
+  // exports React component as default and a 'messages' object for i18n 
+  import Header, { messages } from '@edx/frontend-component-header';
+
+To build a frontend application for a specific brand (edX or other Open edX implementation) we
+will leverage npm aliases to override these npm packages with branded packages that implement the
+same interface before build. For example, ``frontend-component-header`` will be overriden with
+``frontend-component-header-edx``. This is done using the 
+`npm alias syntax introduced in version 6.9.0`_. 
+
+We install aliases using the syntax below:
+
+.. code-block:: bash
+  
+  # npm install <package-name>@<type>:<branded-package>
+  
+  # npm package
+  npm install @edx/frontend-component-header@npm:@edx/frontend-component-header-edx@latest
+
+  # git repository
+  npm install @edx/frontend-component-header@git:https://github.com/edx/frontend-component-header-edx.git
```

Comment:
Out of curiosity, do we know that using `git:` or `file:` causes node_modules to behave the same way as using a published npm package?  I feel like we've seen anecdotal evidence that using filesystem URLs causes npm to act strange... I'm somewhat concerned that we might be sending people down a confusing path.  Whacha think?

Reply:
Don't get me wrong - I think the flexibility is important/it's desirable not to have to use npm.  But I wonder if it's... consistent.

Reply:
Afaik git repos should work nearly identically. For `file:` I know that folders outside the project directory won't get dependencies pulled into the project's node_modules folder like normal. I think this only presents an issue if you're relying on bin scripts to be pulled into the project.